network-config-analyzer/FWRule.py

```diff
@@ -0,0 +1,357 @@
+import ipaddress
+
+from Peer import Pod, IpBlock
+from ConnectionSet import ConnectionSet
+
+
+class LabelExpr:
+    """
+    a class for representing a label expression
+    e.g. key = K, values = {v1, v2} => the label expression is: "K in (v1,v2)"
+    """
+
+    def __init__(self, key, values):
+        """
+        Create a LabelExpr object
+        :param key: a label key of type string
+        :param values: set of label values of type set[string]
+        """
+        self.key = key
+        self.values = values
+
+    def __str__(self):
+        """
+        :return: string representing the label expression
+        """
+        # TODO: handle the string for 'NO_LABEL_VALUE' : app in NO_LABEL_VALUE => !app
+        values_set = set(self.values)
+        values_str = '(' + ','.join(v for v in sorted(list(values_set))) + ')'
+        if ':' not in self.key:
+            return self.key + ' in ' + values_str
+        else:
+            key_labels = self.key.split(':')
+            res = ''
+            values_str_map = dict()
+            # for key in key_labels:
+            for index in range(0, len(key_labels)):
+                key = key_labels[index]
+                split_vals = [val.split(':') for val in self.values]
+                val_set = [v[index] for v in split_vals]
+                values_str = '(' + ','.join(v for v in sorted(list(val_set))) + ')'
+                values_str_map[key] = key + ' in ' + values_str
+            for key in sorted(key_labels):
+                res += values_str_map[key] + ','
+        return res
+
+    def __eq__(self, other):
+        return self.key == other.key and set(self.values) == set(other.values)
```

Comment:
isn't self.values already a set? (that's what the docstring for `__init__` says)

Reply:
yes, changed accordingly